refactor: refactor stub-components.js

lusarz opened the pull request:
Pull request contains changes like
- move public (exported) methods to top of `stub-components.js` file
- separate stub options validation from creation of stubs
- clean a few repeated code
- overall cleaning

lusarz committed:
refactor: refactor stub-components.js

lusarz commented on packages/test-utils/src/shallow.js:
```diff
@@ -26,15 +26,11 @@ export default function shallow (
-      // stubbed components are used instead of original components components
-      ...stubbedGlobalComponents,
-      ...stubbedComponents
+      ...createComponentStubsForGlobals(vue),
```
`stubbedComponents` and `stubbedGlobalComponents` was unnecessary variables, I invoked functions directly

lusarz commented on packages/shared/stub-components.js:
```diff
@@ -37,150 +72,15 @@ function getCoreProperties (component: Component): Object {
-    extended = extended.extends
+function addToIgnoredElements (name) {
+  // ignoreElements does not exist in Vue 2.0.x
+  if (Vue.config.ignoredElements) {
```
These code was duplicated, so I moved it into function

lusarz commented on packages/shared/stub-components.js:
```diff
@@ -37,150 +72,15 @@ function getCoreProperties (component: Component): Object {
-  let extended = component.extends
-
-  // Loop through extended component chains to stub all child components
-  while (extended) {
```
Moved this logic into `createComponentStubsRecursively`

38elements commented:
This looks like only changing style.
Would you tell me the benefits of this?

lusarz commented:
Yes, main purpose of refactoring is changing style and improving the design of an existing code.

I think after these changes code would be easier to maintain and it would be easier to find bugs and introduce extensions.

38elements commented:
Thank you for replying.

There is no objective basis for its benefit.

IMHO,
This is cosmetic change.
I think this change makes difficult to `git blame`. 

This makes difficult to understand code for me.
For example, the `createBlankStub` function helps to understanding code for me,
this function is removed.

lusarz committed:
refactor: refactor stub-components.js - better methods for create stub

lusarz commented:
True, I agree with `git blame` issue. Main reason of that is that I moved public (exported) functions to top of file, what I think is good practice.

I replaced `createBlankStub` with `createStub` method, which returns 'blank' stub by default. But yeah, I agree it was no perfect solution. I updated it in [second commit](https://github.com/vuejs/vue-test-utils/pull/541/commits/25c1adc7e5cdfd800cb0cb3df14ae8d1443be94c). What do you think about that?

38elements commented:
I am sorry.
My expression may have been ambiguous.
It is important whether there is benefit or not.
IMHO,
Since this change has no benefit, I think this change is not necessary.
Other people may not think so.

lusarz committed:
refactor: refactor stub-components.js - replace Object.assign with spread

lusarz commented:
Reducing number of lines of code is beneficial itself

lusarz commented:
I'm not yet happy with everything, so I'm gonna close this pull request for now

eddyerburgh commented:
@38elements this part of the codebase is in need of refactoring, so I think a cosmetic change is fine. If we reduce the lines significantly and keep the code readable, that's a good benefit

38elements commented:
@eddyerburgh 
~~I think that a change which is consistent with the style of existing code is fine.
I think that it is fine if the readability is improved explicitly.~~

38elements commented:
@eddyerburgh 
Since you are a maintainer, 
I think that the code you are easy to maintain is correct.
Would you please do not mind my opinion.
